### PR TITLE
Add flag to include file's dir in logs

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
This PR adds a new flag to `klog` called `add_dir_header`. This flag, when enabled, includes the file's directory in the header.

This flag is meant to fix cases (like in Kubebuilder) where the generated files are called the same across different controllers/folders. This would reduce the need of renaming files, e.g. `machine/controller.go` -> `machine/machine_controller.go`. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: klog.go [hunk 1]
```diff
@@ -585,9 +591,10 @@ func (l *loggingT) header(s severity, depth int) (*buffer, string, int) {
 		file = "???"
 		line = 1
 	} else {
-		slash := strings.LastIndex(file, "/")
-		if slash >= 0 {
-			file = file[slash+1:]
+		dir := path.Base(path.Dir(file))
+		file = path.Base(file)
```
Comment (liggitt) [inline klog.go]: `path.Base(path.Dir(...))` is significantly more expensive than the current code. until we have performance tests in place that tell us whether we are regressing (as several recent functional additions did), I don't think we should merge this

File: klog.go [hunk 2]
```diff
@@ -585,9 +591,10 @@ func (l *loggingT) header(s severity, depth int) (*buffer, string, int) {
...
+		if l.addDirHeader {
+			file = path.Join(dir, file)
```
Comment (liggitt) [inline klog.go]: alternatively, use a second `strings.LastIndex` call to back up a second segment if the addDirHeader option was true, and skip calling path.Base/path.Dir entirely

Comment (dims) [reply klog.go]: @liggitt what happens on windows? :)

Comment (liggitt) [reply klog.go]: what happens today? `/` is hard-coded

Comment (liggitt) [reply klog.go]: also, this is using `path.Base(path.Dir(...))`, not filepath...

Comment (dims) [reply klog.go]: ah good point.

File: klog.go (hunk 1 above)
Comment (vincepri) [reply klog.go]: That's fine. If we can find it useful we can keep it until we have a better solution in place.

File: klog.go (hunk 2 above)
Comment (vincepri) [reply klog.go]: Using a second LastIndex also sounds good. For windows compatibility question, we should probably open another issue

**Comment by liggitt** [comment]
can you add a benchmark that verifies this path does not regress in allocations or CPU when the add dir header option is false?

**Comment by vincepri** [comment]
##  BenchmarkHeader `branch=master`
```bash
$ go test -benchmem -run=^$ k8s.io/klog -bench "^(BenchmarkHeader)$" -v -count=10
goos: darwin
goarch: amd64
pkg: k8s.io/klog
BenchmarkHeader-32    	 2000000	       804 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       802 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       801 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       802 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       803 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       803 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       808 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       815 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       808 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       810 ns/op	     184 B/op	       2 allocs/op
PASS
ok  	k8s.io/klog	24.282s
```

## BenchmarkHeader `branch=directory-header`
```
$ go test -benchmem -run=^$ k8s.io/klog -bench "^(BenchmarkHeader)$" -v -count=10
goos: darwin
goarch: amd64
pkg: k8s.io/klog
BenchmarkHeader-32    	 2000000	       808 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       807 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       807 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       810 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       807 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       807 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       809 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       805 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       806 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeader-32    	 2000000	       807 ns/op	     184 B/op	       2 allocs/op
PASS
ok  	k8s.io/klog	24.345s
```

## BenchmarkHeaderWithDir `branch=directory-header`
```
$ go test -benchmem -run=^$ k8s.io/klog -bench "^(BenchmarkHeaderWithDir)$" -v -count=10
goos: darwin
goarch: amd64
pkg: k8s.io/klog
BenchmarkHeaderWithDir-32    	 2000000	       815 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       814 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       812 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       813 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       815 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       814 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       815 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       814 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       814 ns/op	     184 B/op	       2 allocs/op
BenchmarkHeaderWithDir-32    	 2000000	       813 ns/op	     184 B/op	       2 allocs/op
PASS
ok  	k8s.io/klog	24.556s
```

**Comment by liggitt** [comment]
thanks
/lgtm
/approve

**Comment by dims** [comment]
@vincepri do you mind doc-ing the `go test -benchmem` magic. pretty please? :)

File: klog.go [hunk 3]
```diff
@@ -432,6 +433,7 @@ func InitFlags(flagset *flag.FlagSet) {
 	flagset.BoolVar(&logging.toStderr, "logtostderr", logging.toStderr, "log to standard error instead of files")
 	flagset.BoolVar(&logging.alsoToStderr, "alsologtostderr", logging.alsoToStderr, "log to standard error as well as files")
 	flagset.Var(&logging.verbosity, "v", "number for the log level verbosity")
+	flagset.BoolVar(&logging.skipHeaders, "add_dir_header", logging.addDirHeader, "If true, adds the file directory to the header")
```
Comment (jhahn21) [inline klog.go]: Shouldn't this be `&logging.addDirHeader`?

Comment (liggitt) [reply klog.go]: yes, it should. cc @dims

Comment (dims) [reply klog.go]: @jhahn21 Could you please file a PR to fix it. good catch. thanks!

Comment (jhahn21) [reply klog.go]: @dims Sure, I will.

Comment (jhahn21) [reply klog.go]: @dims Here is the PR: https://github.com/kubernetes/klog/pull/101

**Pull request merged**